clippy_lints/src/clones_into_boxed_slices.rs

```diff
@@ -0,0 +1,204 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::res::MaybeDef;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::sugg::Sugg;
+use clippy_utils::sym;
+use rustc_ast::{BorrowKind, UnOp};
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_middle::ty::{self, Ty};
+use rustc_session::declare_lint_pass;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for clones that are immediately converted into boxed slices instead of using `Box::from(...)`.
+    ///
+    /// ### Why is this bad?
+    /// Using `Box::from(...)` is more concise and avoids creating an unnecessary temporary object.
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let boxed: Box<str> = "example".to_string().into_boxed_str();
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let boxed: Box<str> = Box::from("example");
+    /// ```
+    #[clippy::version = "1.93.0"]
+    pub CLONES_INTO_BOXED_SLICES,
+    perf,
+    "Cloning then converting into boxed slice instead of using Box::from"
+}
+declare_lint_pass!(ClonesIntoBoxedSlices => [CLONES_INTO_BOXED_SLICES]);
+
+fn count_refs(mut expr_ty: Ty<'_>) -> i64 {
+    let mut count = 0;
+    while let ty::Ref(_, inner, _) = expr_ty.kind() {
+        expr_ty = *inner;
+        count += 1;
+    }
+    count
+}
+
+// Shows the lint with a suggestion using the given parts
+// Assures that the inner argument is correctly ref'd/deref'd in the suggestion based on needs_ref
+fn show_lint(
+    cx: &LateContext<'_>,
+    full_span: Span,
+    mut inner: &Expr<'_>,
+    needs_ref: bool,
+    sugg_prefix: Option<&str>,
+    placeholder: &str,
+    sugg_suffix: Option<&str>,
+) {
+    let mut applicability = Applicability::MachineApplicable;
+
+    while let ExprKind::AddrOf(BorrowKind::Ref, _, expr) | ExprKind::Unary(UnOp::Deref, expr) = inner.kind {
+        inner = expr;
+    }
+
+    let mut sugg = Sugg::hir_with_context(cx, inner, full_span.ctxt(), placeholder, &mut applicability);
+
+    let inner_ty = cx.typeck_results().expr_ty(inner);
+    let mut ref_count = count_refs(inner_ty);
```

Comment:
You can use `peel_and_count_ty_refs` from `clippy_utils::ty`:

```suggestion
    let mut ref_count = peel_and_count_ty_refs(inner_ty).1 as isize;
```